Consent container banner

dr3 opened the pull request:
Resolves #1399

**Overall change:** Creates container for cookie and privacy banner

**Code changes:**

- Creates container for cookie and privacy banner
- Adds `/ckns_policy/*` dev route to allow endpoint to use locally.

**Testing:**
1) Check the banner follows the logic described in the description of  https://github.com/bbc/simorgh/issues/1399
2) Check the banner works as expected across devices

Note this will only work on canonical pages e.g.http://localhost:7080/news/articles/cn7769kpk9mo,  not amp pages

Message Drew if you have any questions :)

---

- [x] I have assigned myself to this PR and the corresponding issues
- [x] Tests added for new features
- [ ] Test engineer approval
- [x] I have followed [the merging checklist](https://github.com/bbc/simorgh-infrastructure/blob/latest/documentation/MERGE_PROCESS.md) and this is ready to merge.

jamesbrumpton committed:
Adding Cookie Banner Cypress Tests

jamesbrumpton committed:
Adding Cookie Banner Cypress Tests

Commit:
:rocket: Add consent banner canonical container

Commit:
:rocket: Integrate into Article

Commit:
:rocket: Add dev route for cookie oven

Commit:
:rocket: Add translations config

Commit:
:bath: Eslint fix

Commit:
:bath: Fix naming of logic

pjlee11 commented on src/app/containers/ConsentBanner/Banner.jsx:
```diff
@@ -0,0 +1,63 @@
+const ConsentBannerContainer = ({ type, onReject, onAccept }) => (
+  <ServiceContextConsumer>
+    {({ translations }) => {
+      const messaging = translations.consentBanner[type];
```
```suggestion
      const consentBannerConfig = translations.consentBanner[type];
```

This then means the following lines will be easier to read going from: 
```
title={messaging.title}
reject={Reject(messaging.reject, messaging.rejectUrl, onReject)}
```
to: 
```
title={consentBannerConfig.title}
reject={Reject(consentBannerConfig.reject, consentBannerConfig.rejectUrl, onReject)}
```
which feels easier to understand at a first glance. Also because `messaging.rejectUrl` could be misleading

pjlee11 commented on src/app/containers/ConsentBanner/Banner.jsx:
```diff
@@ -0,0 +1,63 @@
+
+      return (
+        <ConsentBanner
+          title={messaging.title}
```
```suggestion
          title={consentBannerConfig.title}
```

pjlee11 commented on src/app/containers/ConsentBanner/Banner.jsx:
```diff
@@ -0,0 +1,63 @@
+      return (
+        <ConsentBanner
+          title={messaging.title}
+          text={Text(messaging.description)}
```
```suggestion
          text={Text(consentBannerConfig.description)}
```

pjlee11 replied:
```suggestion
          text={Text(consentBannerConfig.description)}
```

pjlee11 commented on src/app/containers/ConsentBanner/Banner.jsx:
```diff
@@ -0,0 +1,63 @@
+        <ConsentBanner
+          title={messaging.title}
+          text={Text(messaging.description)}
+          accept={Accept(messaging.accept, onAccept)}
```
```suggestion
          accept={Accept(consentBannerConfig.accept, onAccept)}
```

pjlee11 commented on src/app/containers/ConsentBanner/Banner.jsx:
```diff
@@ -0,0 +1,63 @@
+          title={messaging.title}
+          text={Text(messaging.description)}
+          accept={Accept(messaging.accept, onAccept)}
+          reject={Reject(messaging.reject, messaging.rejectUrl, onReject)}
```
```suggestion
          reject={Reject(consentBannerConfig.reject, consentBannerConfig.rejectUrl, onReject)}
```

pjlee11 commented on src/app/containers/ConsentBanner/Banner.test.jsx:
```diff
@@ -0,0 +1,20 @@
+  );
+
+  shouldMatchSnapshot(
+    'should correctly render privacy banner',
```
```suggestion
    'should correctly render cookie banner',
```

pjlee11 commented on src/app/containers/ConsentBanner/__snapshots__/index.test.jsx.snap:
```diff
@@ -0,0 +1,9 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
+
+exports[`Consent Banner Container should correctly render amp banner 1`] = `null`;
```
Not sure if I prefer a null snapshot or not having the test for the amp version 🤔

pjlee11 replied:
Actually, this prove the platform context config is correct so lets keep it

pjlee11 commented on src/app/containers/ConsentBanner/index.canonical.test.jsx:
```diff
@@ -0,0 +1,115 @@
+    act(() => {
+      ReactDOM.render(<CanonicalContainer />, container);
+
+      setShowPrivacyBannerState(true);
```
```suggestion
      setShowPrivacyBannerState(true);
      setShowCookieBannerState(false);
```

Would be good to be explicit in this test otherwise the default value of the `setShowCookieBanner` could change the outcome of this test

pjlee11 commented on src/app/containers/ConsentBanner/index.canonical.test.jsx:
```diff
@@ -0,0 +1,115 @@
+      { onAccept: cookieOnAllow, onReject: cookieOnReject, type: 'cookie' },
+      {},
+    );
+  });
```
Think it would be good to have some tests that ensure that when the component is called with both set to false that `logic.not.toHaveBeenCalled` should pass

dr3 replied:
Done :) We now have a test for both set, 1 set, other set, and none set :)

pjlee11 commented on src/app/lib/config/services/news.js:
```diff
@@ -47,6 +47,36 @@ const news = {
+        description: {
+          first:
+            "We've made some important changes to our Privacy and Cookies Policy and we want you to know what this means for you and your data.",
+          linkText: '',
```
```suggestion
          linkText: null,
```

pjlee11 commented on src/app/lib/config/services/news.js:
```diff
@@ -47,6 +47,36 @@ const news = {
+          first:
+            "We've made some important changes to our Privacy and Cookies Policy and we want you to know what this means for you and your data.",
+          linkText: '',
+          last: '',
```
```suggestion
          last: null,
```

pjlee11 commented on src/app/lib/config/services/news.js:
```diff
@@ -47,6 +47,36 @@ const news = {
+            "We've made some important changes to our Privacy and Cookies Policy and we want you to know what this means for you and your data.",
+          linkText: '',
+          last: '',
+          linkUrl: '',
```
```suggestion
          linkUrl: null,
```

pjlee11 commented on src/app/lib/config/services/persian.js:
```diff
@@ -43,6 +43,36 @@ const persian = {
+        description: {
+          first:
+            "We've made some important changes to our Privacy and Cookies Policy and we want you to know what this means for you and your data.",
+          linkText: '',
```
```suggestion
          linkText: null,
```

pjlee11 commented on src/app/lib/config/services/persian.js:
```diff
@@ -43,6 +43,36 @@ const persian = {
+          first:
+            "We've made some important changes to our Privacy and Cookies Policy and we want you to know what this means for you and your data.",
+          linkText: '',
+          last: '',
```
```suggestion
          last: null,
```

pjlee11 commented on src/app/lib/config/services/persian.js:
```diff
@@ -43,6 +43,36 @@ const persian = {
+            "We've made some important changes to our Privacy and Cookies Policy and we want you to know what this means for you and your data.",
+          linkText: '',
+          last: '',
+          linkUrl: '',
```
```suggestion
          linkUrl: null,
```

pjlee11 commented on src/server/index.jsx:
```diff
@@ -79,6 +79,11 @@ server
   .get('/status', (req, res) => {
     res.sendStatus(200);
   })
+  .get('/ckns_policy/*', (req, res) => {
```
Have we explored trying something like the following after the intial `server` config?

```
if (process.env.NODE_ENV === 'development') {
  server
    .get('/ckns_policy/*', (req, res) => {
      res.sendStatus(200);
    });
}
```

Could also be useful for not serving the fixtureData when in production

dr3 replied:
We have, Ive raised https://github.com/bbc/simorgh/issues/1435 to do this in future for all uneeded routes :)

Commit:
Apply suggestions from code review from Phil

Co-Authored-By: dr3 <drew.mcmillan@bbc.co.uk>

Commit:
:bath: Update linting and snapshots for code review

Commit:
:white_check_mark: Add extra test case

ChrisBAshton commented on src/app/containers/ConsentBanner/Banner.jsx:
```diff
@@ -0,0 +1,67 @@
+  linkText: string.isRequired,
+  linkUrl: string.isRequired,
+  last: string.isRequired,
+};
```
Maybe move lines 21-38 out to a `BannerText.jsx` file (and rename to `BannerText`.

dr3 replied:
done with a folder :)

ChrisBAshton commented on src/app/containers/ConsentBanner/index.canonical.jsx:
```diff
@@ -0,0 +1,41 @@
+
+  useEffect(() => {
+    runInitial();
+  }, []);
```
I'm sure you've tried, but is it possible to do:

```js
useEffect(runInitial, []);
```

dr3 replied:
I actually thought I tried this in the past but had an error, I stand corrected, Changed :)

ChrisBAshton commented on src/app/containers/ConsentBanner/index.canonical.test.jsx:
```diff
@@ -0,0 +1,138 @@
+      },
+    );
+
+    CanonicalContainer = require('./index').default;
```
Always a code smell to me - is there a reason this is `.default` whereas line 27 doesn't need it...?

dr3 replied:
Yeah, line 27 exports a function with works, however CanonicalContainer exports a react component which doesnt work without .default sadly

ChrisBAshton commented on src/app/containers/ConsentBanner/index.canonical.test.jsx:
```diff
@@ -0,0 +1,138 @@
+  });
+
+  it('should render privacy banner when both banners are set to be shown', () => {
+    act(() => {
```
_This_ is where I agree with you 😄 about the `it`, `act`, `expect` cycle. Better this than random newlines!

ChrisBAshton commented on src/app/lib/config/services/persian.js:
```diff
@@ -43,6 +43,36 @@ const persian = {
         callToActionLinkUrl: 'https://www.bbc.com/persian',
       },
     },
+    consentBanner: {
```
Let's raise an issue to get Persian translations.

dr3 replied:
Done :) https://github.com/bbc/simorgh/issues/1438

ChrisBAshton reviewed (commented):
Looks good to me. 👍 Added a couple comments you may want to look at.

Commit:
:rocket: Extract banner text component

Commit:
:rocket: Change use effect passing

ChrisBAshton reviewed (dismissed):
![](https://media.giphy.com/media/6CBMidODPz64mNUR9t/giphy.gif)

pjlee11 reviewed (dismissed):
![](https://media3.giphy.com/media/HFHovXXltzS7u/giphy.gif?cid=3640f6095c9100c24e66634345ae9a0c)

jamesbrumpton committed:
Merge branch 'latest' into ConsentContainer_banner

ChrisBAshton committed:
Merge branch 'latest' into ConsentContainer_banner--tests

Commit:
Merge branch 'ConsentContainer_banner' of github.com:BBC-News/simorgh into ConsentContainer_banner--tests

Commit:
Update cypress.json

Commit:
Merge branch 'latest' into ConsentContainer_banner

staylos92 commented:
The logic looks good to me 👍 

Currently performing some device testing:

**Desktop:**
Chrome 72.x 👍 
Firefox 65.x 👍 
Safari 12.x 👍 
Edge 👍 
IE11 👎  - Banner is not displayed and an error in the console says to see "https://09747c45.ngrok.io/news/articles/cn7769kpk9mo" for more info.


**Mobile:**
Chrome 72 on Android 👍 
Firefox 65 on Android 👍 
Safari on iOS 12 👍 
Safari on iOS 11 👍 
Safari on iOS 10 👍 
Safari iOS9 👍

ChrisBAshton committed:
adds additional cookie edge-case test

ChrisBAshton committed:
fixes indentation of cypress.json file

ChrisBAshton committed:
Merge branch 'latest' into ConsentContainer_banner--tests

ChrisBAshton committed:
_actually_ fixes indentation in cypress.json

Something went wrong in https://github.com/bbc/simorgh/pull/1439/commits/b6bd1507be4b752a665685d834d641a358e999ed

ChrisBAshton committed:
Merge branch 'latest' into ConsentContainer_banner

ChrisBAshton committed:
Merge branch 'ConsentContainer_banner' into ConsentContainer_banner--tests

ChrisBAshton committed:
renames retrieveCookieValue to assertCookieValue

Makes intent clearer

ChrisBAshton committed:
refactor cookie expiry functions according to code review

ChrisBAshton committed:
namespace cy.contains() calls and refactors one test

ChrisBAshton committed:
assert default cookie policy not overwritten on page load

ChrisBAshton committed:
refactor tests

ChrisBAshton commented:
Blocked pending #1439, but otherwise ready for merge.

We've not explicitly been able to check it in IE but will fix forward.

ChrisBAshton committed:
adds test for returning users who haven't explicitly accepted cookies

ChrisBAshton committed:
prefer double-quotes than escaping

Commit:
rename test more descriptively

Co-Authored-By: ChrisBAshton <ChrisBAshton@users.noreply.github.com>

ChrisBAshton committed:
refactor tests - make explicit when visiting article and setting cookies

ChrisBAshton committed:
Merge branch 'latest' into ConsentContainer_banner--tests

ChrisBAshton committed:
merge with ConsentContainer_banner--tests on remote

ChrisBAshton committed:
Merge branch 'latest' into ConsentContainer_banner

ChrisBAshton committed:
increase buffer to 15 milliseconds to allow for slow tests

ChrisBAshton committed:
Merge pull request #1439 from bbc/ConsentContainer_banner--tests

Consent container banner  tests

ChrisBAshton committed:
increases buffer to 20 milliseconds

dr3 reviewed (commented):
👍 I cant officially approve but 👍

pjlee11 reviewed (approved):
LGTM 👍
